src/pipeline.ts

```diff
@@ -335,28 +339,108 @@ export async function checkNewCandidateMatch(
+			metas
+				.map(createSearcheeFromMetafile)
+				.filter(isOk)
+				.map((r) => r.unwrap())
+				.map((searchee) => ({ ...searchee, label: searcheeLabel }))
+				.filter((searchee) => filterByContent(searchee)),
+		),
 	);
 	if (!searchees.length) {
 		logger.verbose({
 			label: searcheeLabel,
 			message: `No valid entries found using ${method} for ${candidateLog}`,
 		});
-		return { decision: null, actionResult: null };
+		return null;
+	}
+	return method;
+}
+
+async function pushEnsembleForCandidate(
+	candidate: Candidate,
+	searchees: SearcheeWithLabel[],
+	searcheeLabel: SearcheeLabel,
+): Promise<string | null> {
+	const { seasonFromEpisodes } = getRuntimeConfig();
+	if (!seasonFromEpisodes) return null;
+	const seasonKey = getSeasonKey(stripExtension(candidate.name));
+	if (!seasonKey) return null;
+
+	const candidateLog = `${chalk.bold.white(candidate.name)} from ${candidate.tracker}`;
+	const { ensembleTitle, keyTitle, season } = seasonKey;
+	const key = `${keyTitle}.${season}`;
+	const ensemble = await memDB("ensemble").where({ ensemble: key });
+	if (ensemble.length === 0) {
+		logger.verbose({
+			label: searcheeLabel,
+			message: `Did not find an ensemble ${ensembleTitle} for ${candidateLog}`,
+		});
+		return null;
+	}
+	const files = ensemble
+		.map((e) => ({
+			path: e.absolute_path,
+			name: e.name,
+			length: e.length,
+		}))
+		.filter((f) => fs.existsSync(f.path));
+	if (files.length === 0) {
+		logger.verbose({
+			label: searcheeLabel,
+			message: `Did not find any files for ensemble ${ensembleTitle} for ${candidateLog}: sources may be incomplete or missing`,
+		});
+		return null;
 	}
+	// Get searchee.length by total size of each episode (average if multiple files for episode)
+	const uniqueElements = new Set(ensemble.map((e) => e.element));
+	const totalLength = [...uniqueElements].reduce((acc, cur) => {
+		const elements = ensemble.filter(
+			(e) =>
+				e.element === cur &&
+				files.some((f) => f.path === e.absolute_path),
+		);
+		const avg =
+			elements.reduce((a, c) => a + c.length, 0) / elements.length;
+		return acc + avg;
+	}, 0);
+	searchees.push({
+		name: ensembleTitle,
+		title: ensembleTitle,
+		files: files,
+		length: totalLength,
+		mtimeMs: await getNewestFileAge(files.map((f) => f.path)),
+		label: searcheeLabel,
+	});
 	logger.verbose({
 		label: searcheeLabel,
-		message: `Unique entries [${searchees.map((m) => m.title)}] using ${method} for ${candidateLog}`,
+		message: `Using ensemble ${ensembleTitle} for ${candidateLog}: ${humanReadableSize(totalLength)} - ${files.length} files`,
 	});
+	return "ensemble";
+}
 
+export async function checkNewCandidateMatch(
+	candidate: Candidate,
+	searcheeLabel: SearcheeLabel,
+): Promise<{
+	decision: DecisionAnyMatch | Decision.INFO_HASH_ALREADY_EXISTS | null;
+	actionResult: ActionResult | null;
+}> {
+	const searchees: SearcheeWithLabel[] = [];
+
+	const methods = [
+		await pushSearcheesForCandidate(candidate, searchees, searcheeLabel),
```

Comment:
is there a hard reason we need to push into this array of searchees rather than having these functions return arrays themselves and concatenating them together into a new array? I find it obfuscates the data flow especially when the creation and mutation are in separate functions.

Reply:
I needed to return the string for logging, so I felt pushing to the array makes it easier. The alternative would be a more compilated return type.

Reply:
Let's do the more complicated return type.